node-api/src/main/kotlin/net/corda/nodeapi/internal/lifecycle/NodeLifecycleEventsDistributor.kt

```
@@ -0,0 +1,109 @@
+                NodeLifecycleEvent.BeforeStart::class.java,
+                NodeLifecycleEvent.AfterStart::class.java)
+        private val criticalExceptionsClasses: Set<Class<out Throwable>> = setOf(CordaServiceCriticalFailureException::class.java)
+    }
+
+    /**
+     * Order is maintained by priority and within equal priority by full class name.
+     */
+    private val prioritizedObservers: MutableList<NodeLifecycleObserver> = mutableListOf()
+
+    private val readWriteLock: ReadWriteLock = ReentrantReadWriteLock()
+
+    /**
+     * Adds observer to the distribution list.
+     */
+    fun <T : NodeLifecycleObserver> add(observer: T) : T {
+        addAll(singleton(observer))
+        return observer
+    }
+
+    /**
+     * Adds multiple observers to the distribution list.
+     */
+    fun <T : NodeLifecycleObserver> addAll(observers: Collection<T>) : Collection<T> {
+
+        data class SortingKey(val priority: Int, val clazz: Class<*>) : Comparable<SortingKey> {
+            override fun compareTo(other: SortingKey): Int {
+                return if(priority != other.priority) {
+                    priority - other.priority
+                } else {
+                    clazz.name.compareTo(other.clazz.name)
+                }
+            }
+        }
+
+        readWriteLock.writeLock().executeLocked {
+            prioritizedObservers.addAll(observers)
+            // In-place sorting
+            prioritizedObservers.sortBy { SortingKey(it.priority, it.javaClass) }
+        }
+
+        return observers
+    }
+
+    /**
+     * Distributes event to all the observers previously added
+     */
+    fun distributeEvent(event: NodeLifecycleEvent) {
+        val snapshot = readWriteLock.readLock().executeLocked { LinkedList(prioritizedObservers) }
+        val orderedSnapshot = if (event.reversedPriority) snapshot.reversed() else snapshot
+
+        val updateResult = orderedSnapshot.map { it.update(event) }
```

I'm not sure it makes sense to carry on in the event of a failure.  We should fail fast, rather than analysing at the end.  Who knows what harm can be caused by continuing partially initialised.

---

OK, I will change this logic.